iroh-net/src/discovery/static_provider.rs

```diff
@@ -0,0 +1,162 @@
+//! A static discovery implementation that allows adding info for nodes manually.
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    sync::{Arc, RwLock},
+    time::SystemTime,
+};
+
+use futures_lite::stream::{self, StreamExt};
+use iroh_base::{
+    key::NodeId,
+    node_addr::{AddrInfo, NodeAddr},
+};
+
+use super::{Discovery, DiscoveryItem};
+
+/// A static discovery implementation that allows providing info for nodes manually.
+#[derive(Debug, Default)]
+#[repr(transparent)]
+pub struct StaticProvider {
+    nodes: Arc<RwLock<BTreeMap<NodeId, NodeInfo>>>,
+}
+
+#[derive(Debug)]
+struct NodeInfo {
+    info: AddrInfo,
+    last_updated: SystemTime,
+}
+
+impl StaticProvider {
+    /// The provenance string for this discovery implementation.
```

Comment:
This should say what it is for.  Like this it is completely lost on the user.

e.g. local swarm discovery uses this text:
```
/// Used as the `provenance` field in [`DiscoveryItem`]s.
///
/// Used in the [`crate::endpoint::Source::Discovery`] enum variant as the `name`.
```

I like having it as an associated constant though :)